api/service/legacysync/syncing.go

```diff
@@ -822,16 +848,19 @@ func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker
 	parentHash := bc.CurrentBlock().Hash()
 
 	var err error
-	for {
-		block := ss.getBlockFromOldBlocksByParentHash(parentHash)
+	for block := ss.getBlockFromOldBlocksByParentHash(parentHash); ; {
 		if block == nil {
 			break
 		}
-		err = ss.UpdateBlockAndStatus(block, bc, worker, false)
+		next := ss.getBlockFromOldBlocksByParentHash(block.Hash())
+		// Enforce sig check for the last block in a batch
+		enforceSigCheck := next == nil
+
+		err = ss.UpdateBlockAndStatus(block, bc, worker, enforceSigCheck)
 		if err != nil {
 			break
 		}
-		parentHash = block.Hash()
+		block = next
```

Comment:
parentHash is not updated. This loop doesn't seem to work anymore.

Reply:
ParentHash is now only used once. For the rest of the loop, `block.Hash()` is used and keeps iterating.

Reply:
I see. this logic seems more convoluted than the previous one. Why change the loop logic? The previous one seems more logical for readers.

Reply:
Because I want to check whether the current block is the last one in `commonBlocks`. Sorry for not focusing on the code quality of these codes since I thought the code was about to be discarded. Let me try refactor a little bit.

Reply:
Just refactored. Please let me know whether the new logic makes sense to you. Thanks for pointing this out!